Apps/JobScheduler/src/Jobs/CommunicationJob.cs

```diff
@@ -0,0 +1,147 @@
+
+                foreach (Communication comm in communications)
+                {
+#pragma warning disable CA1031 //We want to catch exception.
+                    try
+                    {
+                        DateTime? createdOnOrAfterFilter = null;
+                        string lastProcessedProfileHdid = string.Empty;
+                        List<UserProfile>? usersToSendCommEmails = null;
+                        do
+                        {
+                            if (usersToSendCommEmails != null && usersToSendCommEmails.Count > 0)
+                            {
+                                lastProcessedProfileHdid = usersToSendCommEmails[usersToSendCommEmails.Count - 1].HdId;
+                            }
+
+                            usersToSendCommEmails = this.commEmailDelegate.GetActiveUserProfilesWithoutCommEmailByCommunicationIdAndByCreatedOnOrAfter(comm.Id, createdOnOrAfterFilter, this.retryFetchSize);
+                            foreach (UserProfile profile in usersToSendCommEmails)
+                            {
+                                // Idea for Improvement: we can look into a lib for Bulk Insert / Update using .NET Core for Postgres (if needed).
+                                // Insert a new Email record into db.
+                                Email email = new Email()
+                                {
+                                    From = this.fromEmailAddressHGDonotreply,
+                                    To = profile.Email,
+                                    Subject = comm.Subject,
+                                    Body = comm.Text,
+                                    FormatCode = EmailFormat.HTML,
+                                    Priority = comm.Priority,
+                                };
+                                this.emailQueueService.QueueNewEmail(email, false);
+
+                                // Inser a new CommunicationEmail record into
+                                CommunicationEmail commEmail = new CommunicationEmail()
+                                {
+                                    Email = email,
+                                    UserProfile = profile,
+                                    Communication = comm,
+                                };
+                                this.commEmailDelegate.Add(commEmail, false);
+                            }
+
+                            if (usersToSendCommEmails.Count > 0)
+                            {
+                                // Set the createdOnOrAfterFilter for the next query retrieving the next chunk of user profiles.
+                                createdOnOrAfterFilter = usersToSendCommEmails[usersToSendCommEmails.Count - 1].CreatedDateTime;
+                            }
+
+                            this.dbContext.SaveChanges(); // commit after every retryFetchSize (or 250) pairs of Email & CommunicationEmail.
```

Comment:
Now that I see this here, I see that we’re missing a state... we should flip the state to processing or something similar when we start and the outer query should look for new and processing’s comms.

Reply:
I will add a new "Processing" status to the Communication's Communication Status Code and update it when processing a communication in the job scheduler.

Reply:
Code updated per @sslaws 's great comments above.